manifests/server/backup.pp

```diff
@@ -1,62 +1,43 @@
 # See README.me for usage.
 class mysql::server::backup (
-  $backupuser,
-  $backuppassword,
-  $backupdir,
-  $backupdirmode = '0700',
-  $backupdirowner = 'root',
-  $backupdirgroup = 'root',
-  $backupcompress = true,
-  $backuprotate = 30,
-  $ignore_events = true,
+  $backupuser         = undef,
+  $backuppassword     = undef,
+  $backupdir          = undef,
+  $backupdirmode      = '0700',
+  $backupdirowner     = 'root',
+  $backupdirgroup     = 'root',
+  $backupcompress     = true,
+  $backuprotate       = 30,
+  $ignore_events      = true,
   $delete_before_dump = false,
-  $backupdatabases = [],
-  $file_per_database = false,
-  $ensure = 'present',
-  $time = ['23', '5'],
-  $postscript = false,
-  $execpath   = '/usr/bin:/usr/sbin:/bin:/sbin',
+  $backupdatabases    = [],
+  $file_per_database  = false,
+  $ensure             = 'present',
+  $time               = ['23', '5'],
+  $postscript         = false,
+  $execpath           = '/usr/bin:/usr/sbin:/bin:/sbin',
+  $provider           = 'mysqldump',
 ) {
 
-  mysql_user { "${backupuser}@localhost":
-    ensure        => $ensure,
-    password_hash => mysql_password($backuppassword),
-    provider      => 'mysql',
-    require       => Class['mysql::server::root_password'],
-  }
-
-  mysql_grant { "${backupuser}@localhost/*.*":
-    ensure     => $ensure,
-    user       => "${backupuser}@localhost",
-    table      => '*.*',
-    privileges => [ 'SELECT', 'RELOAD', 'LOCK TABLES', 'SHOW VIEW', 'PROCESS' ],
-    require    => Mysql_user["${backupuser}@localhost"],
-  }
-
-  cron { 'mysql-backup':
-    ensure  => $ensure,
-    command => '/usr/local/sbin/mysqlbackup.sh',
-    user    => 'root',
-    hour    => $time[0],
-    minute  => $time[1],
-    require => File['mysqlbackup.sh'],
-  }
-
-  file { 'mysqlbackup.sh':
-    ensure  => $ensure,
-    path    => '/usr/local/sbin/mysqlbackup.sh',
-    mode    => '0700',
-    owner   => 'root',
-    group   => 'root',
-    content => template('mysql/mysqlbackup.sh.erb'),
-  }
-
-  file { 'mysqlbackupdir':
-    ensure => 'directory',
-    path   => $backupdir,
-    mode   => $backupdirmode,
-    owner  => $backupdirowner,
-    group  => $backupdirgroup,
-  }
+  create_resources("mysql::backup::${provider}", {
```

Comment:
Could you convert the backup defines to classes (since they will only ever be used once) and have this create_resources of type "class")?